### ISSUE TITLE
[-Wnullable-to-nonnull-conversion] is weak and dumb

### ISSUE THREAD

**Issue by alejandro-colomar** [llvm/llvm-project#57546]
I was trying the feature, but it is trivial to silence warnings, and it also warns in some cases where it shouldn't:

```c
$ cat nonnull.c 
int *_Nonnull f1(int *_Nullable p)
{
	return p;  // We got a warning :)
}

int *_Nonnull f2(int *_Nullable p)
{
	int *_Null_unspecified q;

	q = p;  // Expected a warning, but not :/

	return q;  // Expected a warning, but not :/
}

int f3(int *_Nullable p)
{
	return *p;  // Expected a warning, but not :/
}

int *_Nonnull f4(int *_Nullable p, int *_Nonnull q)
{
	if (p)
		return p;  // Should not get a warning here :(
	return q;
}
```
```sh
$ clang nonnull.c -Weverything -Wno-nullability-extension -Wno-missing-prototypes -S -o /dev/null
nonnull.c:3:9: warning: implicit conversion from nullable pointer 'int * _Nullable' to non-nullable pointer type 'int * _Nonnull' [-Wnullable-to-nonnull-conversion]
        return p;  // We got a warning :)
               ^
nonnull.c:23:10: warning: implicit conversion from nullable pointer 'int * _Nullable' to non-nullable pointer type 'int * _Nonnull' [-Wnullable-to-nonnull-conversion]
                return p;  // Should not get a warning here :(
                       ^
2 warnings generated.
```

**Comment by tbaederr** [comment]
Could you please change the title to something that could be considered professional and constructive instead of "weak and dumb"?

**Comment by alejandro-colomar** [comment]
> Could you please change the title to something that could be considered professional and constructive instead of "weak and dumb"?

Sure.  Done.  Sorry I was a bit unimaginative for that yesterday. :)

**Comment by MosheBerman** [comment]
## tldr

1. I think this is `f3` expected behavior because of `null_unspecified` and also the checker tries hard to avoid noise. 

2. `f4` can be relatively easily implemented in the static analyzer but I'm not sure about the Semantic Type system. 

3. Happy to (tentatively) look at implementing a fix for `f4`, time permitting. 

## In Detail

@dwblaikie, the page you linked says that dereferences are supposed to be handled *when they are annotated*. My reading is that this is `f3` is expected behavior. 

Put differently,the checker only checks for `nullable` and `nonnull` conversions, not `nullable` to `null_unspecified` conversions. 

`null_unspecified` is the default annotation, and `f3` has an unannotated return type. 

Similar to the page you linked, [this page](https://discourse.llvm.org/t/rfc-nullability-qualifiers/35672) has the description and the ensuing discussion. 

The source is in NullabilityChecker.cpp and there are some useful comments which address the design choices. (I'll try to grab a link - hard to do that when I'm commenting from my phone.)

There is also some logic in [SemaType.cpp](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/Sema.cpp), around local variables and declaration shadowing. This is where I suspect the complexity around f4 comes in. 

> Source: I worked on an automatic nullability annotator at Meta last year and spent a ton of time in this code. ([Here's an early version](https://reviews.llvm.org/D123352) that I proposed before mass layoffs.)

**Comment by alejandro-colomar** [comment]
Hi!

> ### tldr
> 
>     1. I think this is `f3` expected behavior because of `null_unspecified` and also the checker tries hard to avoid noise.

Seems reasonable, especially since you have the `#pragmas` to tweak the default.

> 
>     2. `f4` can be relatively easily implemented in the static analyzer but I'm not sure about the Semantic analyzer.

Hmm, since this warning is out of -Wextra, then it also seems reasonable.

BTW, I'd like to ask if there's a way to run the static analyzer as easily as GCC's `-fanalyzer`.  From what I see, Clang's analyzer is something that takes over your build system and does black magic to it to guess the right thing to do.  I'd like instead something very dumb to which I can tell exactly what to do.

Is there some command that I can run on files, as `clang-analyzer $CLANG-ANALYZERFLAGS $CFLAGS foo.c` and get warnings out of it, similar to how `clang-tidy` works?

Should I open an issue to request this feature?

> 
>     3. Happy to (tentatively) look at implementing a fix for `f4`, time permitting.

I'd love that.

> 
> 
> ### In Detail
> 
> @dwblaikie, the page you linked says that dereferences are supposed to be handled _when they are annotated_. My reading is that this is `f3` is expected behavior.
> 
> Put differently,the checker only checks for `nullable` and `nonnull` conversions, not `nullable` to `null_unspecified` conversions.
> 
> `null_unspecified` is the default annotation, and `f3` has an unannotated return type.
> 
> Similar to the page you linked, [this page](https://discourse.llvm.org/t/rfc-nullability-qualifiers/35672) has the description and the ensuing discussion.
> 
> The source is in [NullabilityChecker.cpp](https://github.com/llvm/llvm-project/blob/main/clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp) and there are some useful comments which address the design choices.
> 
> There is also some logic in [SemaType.cpp](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/Sema.cpp), around local variables and declaration shadowing. This is where I suspect the complexity around f4 comes in.
> 
> > Source: I worked on an automatic nullability annotator at Meta last year and spent a ton of time in this code. (Just so you know I'm not making that up, [here's an early version](https://reviews.llvm.org/D123352) that I proposed before mass layoffs ended my work on that. The proposal was reject as-it-was but I had a nicer  clang-tidy version in the works. C'est La Vie.)

Thanks a lot!

Alex

**Comment by MosheBerman** [comment]
> BTW, I'd like to ask if there's a way to run the static analyzer as easily as GCC's -fanalyzer. 

Yes, but I don't remember offhand. We had these build commands saved to build config files, so my muscle memory is lacking here.  

Let me look it up. 

> From what I see, Clang's static analyzer is something that takes over your build system and does black magic to it to guess the right thing to do. I'd like instead something very dumb to which I can tell exactly what to do.

My limited understanding of the dark "magic" is that it traverses a giant syntax tree. (as part of the compilation, it parses the code into an Abstract Syntax Tree aka AST.) Simplistically, static analyzers are implemented as tree-traversal functions. 

> Is there some command that I can run on files, as clang-analyzer $CLANG-ANALYZERFLAGS $CFLAGS foo.c and get warnings out of it, similar to how clang-tidy works?

Yes, same as above. 

> Should I open an issue to request this feature?

I'm not a maintainer/collaborator, so grain of salt, but in my opinion this issue can be used. Let's see what the others think.

**Comment by MosheBerman** [comment]
🤦‍♂️right, should be simple. 

You can add `-Wnullability` to your regular build command for static nullability warnings. 

The list of warnings [is here](https://clang.llvm.org/docs/DiagnosticsReference.html#id583)

**Comment by alejandro-colomar** [comment]
> man_facepalmingright, should be simple.
> 
> You can add `-Wnullability` to your regular build command for static nullability warnings.
> 
> The list of warnings [is here](https://clang.llvm.org/docs/DiagnosticsReference.html#id583)

No what I meant with "it's not in `-Wextra`", is that I like that it's not enabled by default.  Since it has false positives, like `f4`, it would be quite bad if it was in `-Wextra`, but since right now it's just opt-in, it's a good thing.

**Comment by alejandro-colomar** [comment]
I would like some static analyzer that has no false positives, like hopefully <https://clang-analyzer.llvm.org/>, to be able to run it in my build system.

If you can find the commands to run it as a simple command on files, without telling it about my build system, it would be great!

**Comment by nickdesaulniers** [comment]
This came up today on LKML: https://lore.kernel.org/lkml/CAHk-=wiP0983VQYvhgJQgvk-VOwSfwNQUiy5RLr_ipz8tbaK4Q@mail.gmail.com/

Personally, I was surprised that `_Nonnull` on return types didn't seem to do anything:
```c
#include <stddef.h>

int x;

int * _Nonnull foo (void) {
    return &x;
}

void baz (int * _Nonnull);

void bar (void) {
    if (foo() == NULL) // maybe should warn that foo() returns _Nonnull?
        bar();
}
```
especially since android is adding these to bionic.
https://android-review.googlesource.com/c/platform/bionic/+/2552567/7/libc/include/netinet/ether.h#45
https://android-review.googlesource.com/q/owner:zijunzhao@google.com+Nullability

**Comment by alejandro-colomar** [comment]
> ### tldr
> 
>     1. I think this is `f3` expected behavior because of `null_unspecified` and also the checker tries hard to avoid noise.


> `null_unspecified` is the default annotation, and `f3` has an unannotated return type.

I have realized right now after re-reading this bug report that these comments about f3 are not correct.  The problem in f3 is that it is dereferencing a nullable pointer.  It is returning an int, so the return type is not relevant.

@MosheBerman
